File: InsertionSort.ts
function insertion_sort(arr: number[]): number[] {
    for (let i = 1; i < arr.length; i++) {
        let key = arr[i];
        let j = i - 1;
        while (j >= 0 && key < arr[j]) {
            arr[j + 1] = arr[j]
            j--
        }
        arr[j + 1] = key;
    }
    return arr;
}

function insertion_sort_reverse(arr: number[]): number[] {
    for (let i = 1; i < arr.length; i++) {
        let key = arr[i]
        let j = i - 1;
        while (j >= 0 && key > arr[j]) {
            arr[j + 1] = arr[j]
            j--
        }
        arr[j + 1] = key;
    }
    return arr;
}

let arr = [4, 2, 0, 6, 9]

let ascending = insertion_sort(arr)
console.log(ascending)

let descending = insertion_sort_reverse(arr)
console.log(descending)

File: InsertionSort.go
package main

import "fmt"

func insertionSort(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1
		for j >= 0 && key < arr[j] {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
	return arr
}

func insertionSortReverse(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1
		for j >= 0 && key > arr[j] {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
	return arr
}

func main() {
	arr := []int{4, 2, 0, 6, 9}

	ascending := insertionSort(arr)
	fmt.Printf("%v", ascending)

    fmt.Println();

	descending := insertionSortReverse(arr)
	fmt.Printf("%v", descending)
}
